base/applications/taskmgr/graphctl.c

```diff
@@ -12,666 +12,363 @@
+    INT     p;
+    RECT    rc;
+
+    inst->hParentWnd = hParentWnd;
+    inst->hWnd = hWnd;
+
+    Size = GetSystemMetrics(SM_CXSCREEN);
+    if (!Size)
+        Size = 1920;
+    inst->BitmapWidth = Size;
+    Size /= PLOT_SHIFT;
+    t = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, Size * NUM_PLOTS);
+    inst->PointBuffer = t;
+    if (!t)
     {
-        this->m_penPlot[i] = CreatePen(PS_SOLID, 0, this->m_crPlotColor[i]);
+        PostQuitMessage(0);
     }
-    this->m_brushBack = CreateSolidBrush(this->m_crBackColor);
-
-    /*  public member variables, can be set directly  */
-    strcpy(this->m_strXUnitsString, "Samples");  /*  can also be set with SetXUnits */
-    strcpy(this->m_strYUnitsString, "Y units");  /*  can also be set with SetYUnits */
-
-    /*  protected bitmaps to restore the memory DC's */
-    this->m_bitmapOldGrid = NULL;
-    this->m_bitmapOldPlot = NULL;
-}
-
-void GraphCtrl_Dispose(TGraphCtrl* this)
-{
-    int plot;
-
-    for (plot = 0; plot < MAX_PLOTS; plot++)
-        DeleteObject(this->m_penPlot[plot]);
-
-    /*  just to be picky restore the bitmaps for the two memory dc's */
-    /*  (these dc's are being destroyed so there shouldn't be any leaks) */
-
-    if (this->m_bitmapOldGrid != NULL) SelectObject(this->m_dcGrid, this->m_bitmapOldGrid);
-    if (this->m_bitmapOldPlot != NULL) SelectObject(this->m_dcPlot, this->m_bitmapOldPlot);
-    if (this->m_bitmapGrid    != NULL) DeleteObject(this->m_bitmapGrid);
-    if (this->m_bitmapPlot    != NULL) DeleteObject(this->m_bitmapPlot);
-    if (this->m_dcGrid        != NULL) DeleteDC(this->m_dcGrid);
-    if (this->m_dcPlot        != NULL) DeleteDC(this->m_dcPlot);
-    if (this->m_brushBack     != NULL) DeleteObject(this->m_brushBack);
-}
-
-void GraphCtrl_Create(TGraphCtrl* this, HWND hWnd, HWND hParentWnd, UINT nID)
-{
-    GraphCtrl_Init(this);
-    this->m_hParentWnd = hParentWnd;
-    this->m_hWnd = hWnd;
-
-    GraphCtrl_Resize(this);
+    inst->NumberOfPoints = Size;
+    inst->CurrIndex = 0;
+
+    /* styling */
+    inst->hPenGrid = CreatePen(PS_SOLID, 0, fmt->clrGrid);
+    inst->hPen0 = CreatePen(PS_SOLID, 0, fmt->clrPlot0);
+    inst->hPen1 = CreatePen(PS_SOLID, 0, fmt->clrPlot1);
+    inst->hBrushBack = CreateSolidBrush(fmt->clrBack);
+
+    if (fmt->GridCellWidth >= PLOT_SHIFT << 2)
+        inst->GridCellWidth = fmt->GridCellWidth;
+    else
+        inst->GridCellWidth = PLOT_SHIFT << 2;
+    if (fmt->GridCellHeight >= PLOT_SHIFT << 2)
+        inst->GridCellHeight = fmt->GridCellHeight;
+    else
+        inst->GridCellHeight = PLOT_SHIFT << 2;
+
+    inst->DrawSecondaryPlot = fmt->DrawSecondaryPlot;
+
+    GetClientRect(hWnd, &rc);
+    rc.bottom++;
+    inst->BitmapHeight = rc.bottom;
+    inst->ftPixelsPerPercent = (FLOAT)(inst->BitmapHeight) / 100.00f;
+
+    hdc = GetDC(hParentWnd);
+    hdcg = CreateCompatibleDC(hdc);
+    inst->hdcGraph = hdcg;
+    inst->hbmGraph = CreateCompatibleBitmap(hdc, inst->BitmapWidth, inst->BitmapHeight);
```

Comment:
You may or may not check whether the GetDC/CreateCompatibleDC/CreateCompatibleBitmap succeed, and return FALSE in case they don't (and do corresponding cleanups).
Same comment as well for the CreatePen/CreateSolidBrush calls from above.